src/java/org/apache/cassandra/index/sai/plan/StorageAttachedIndexSearcher.java

```diff
@@ -403,10 +403,13 @@ public void close()
 
         public UnfilteredRowIterator apply(PrimaryKey key)
         {
-            // Key reads are lazy, delayed all the way to this point. Skip if we've already seen this one:
-            if (key.equals(lastKey))
+            // Key reads are lazy, delayed all the way to this point.
+            // We don't want key.equals(lastKey) because some PrimaryKey implementations consider more than just
+            // partition key and clustering for equality. This can break lastKey skipping, which is necessary for
+            // correctness when PrimaryKey doesn't have a clustering (as otherwise, the same partition may get
+            // filtered and considered as a result multiple times).
+            if (lastKey != null && key.partitionKey().equals(lastKey.partitionKey()) && key.clustering().equals(lastKey.clustering()))
```

Comment:
issue: `key.clustering()` can return null if the key was created with `RowAwarePrimaryKeyFactory.createTokenOnly(token)`.